src/wallet/rpc/wallet.cpp

```diff
@@ -437,6 +438,14 @@ static RPCHelpMan unloadwallet()
         throw JSONRPCError(RPC_WALLET_NOT_FOUND, "Requested wallet does not exist or is not loaded");
     }
 
+    if (wallet->IsScanning()) {
+        if (request.params[2].isNull() || !request.params[2].get_bool()) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot unload a wallet while it is rescanning. Retry with the force option or run abortrescan first.");
+        }
+
+        wallet->AbortRescan();
```

Comment:
As `AbortRescan()` does not abort the scanning process right away (only sets a flag that is later read by the wallet's scanning thread), might be good to add a "wait until scan aborts" here.

So there is no race between the wallet removal and the scanning process.